mantis-connectors/mantis-connector-iceberg/src/test/java/io/mantisrx/connector/iceberg/sink/writer/IcebergWriterEndToEndTest.java

```diff
@@ -0,0 +1,166 @@
+          "host");
+
+  private Partitioner partitioner = record -> {
+    GenericRecord partitionRecord = GenericRecord.create(spec.schema());
+    partitionRecord.setField("partition", 1);
+    return partitionRecord;
+  };
+
+  private Context stageContext = mock(Context.class);
+
+  @BeforeEach
+  public void initStageContext() {
+    when(stageContext.getWorkerInfo()).thenReturn(workerInfo);
+  }
+
+  @Test
+  public void testWhenMultipleThreadsAreUsedForWritingToTheTable() throws Exception {
+    final WriterConfig writerConfig = new WriterConfig(
+        tableExtension.getCatalog(),
+        tableExtension.getDatabase(),
+        tableExtension.getTableName(),
+        WRITER_ROW_GROUP_SIZE_DEFAULT,
+        Long.parseLong(WRITER_FLUSH_FREQUENCY_BYTES_DEFAULT),
+        100L,
+        WRITER_FILE_FORMAT_DEFAULT,
+        WRITER_MAXIMUM_POOL_SIZE_DEFAULT,
+        new Configuration());
+
+    final IcebergWriterFactory writerFactory =
+        new DefaultIcebergWriterFactory(writerConfig, workerInfo, tableExtension.getTable(),
+            tableExtension.getLocationProvider());
+
+    final IcebergWriterPool writerPool = new FixedIcebergWriterPool(writerFactory, writerConfig);
+    Transformer transformer =
+        IcebergWriterStage.newTransformer(writerConfig, new WriterMetrics(), writerPool, partitioner, stageContext);
+
+    final int size = 1000;
+
+    // odd numbers observable
+    Observable<GenericRecord> oddObservable =
+        Observable
+            .interval(0, 1, TimeUnit.MILLISECONDS)
+            .filter(val -> (val % 2 == 1))
+            .takeUntil(val -> val > size)
+            .map(val -> {
+              GenericRecord record = GenericRecord.create(schema);
+              record.setField("partition", 1);
+              record.setField("id", val);
+              return record;
+            })
+            .publish()
+            .autoConnect();
+
+    // even numbers observable
+    Observable<GenericRecord> evenObservable =
+        Observable
+            .interval(size / 2, 1, TimeUnit.MILLISECONDS)
+            .filter(val -> (val % 2 == 0))
+            .takeUntil(val -> val + (size /2) > size)
+            .map(val -> {
+              GenericRecord record = GenericRecord.create(schema);
+              record.setField("partition", 1);
+              record.setField("id", val);
+              return record;
+            })
+            .publish()
+            .autoConnect();
+
+    Observable<Record> recordObservable = Observable.merge(oddObservable, evenObservable);
+
+    Observable<DataFile> dataFileObservable = transformer.call(recordObservable);
+
+    AtomicReference<Throwable> failure = new AtomicReference<>();
+    List<DataFile> dataFileList = new ArrayList<>();
+    Subscription subscription = dataFileObservable.subscribe(dataFileList::add, failure::set);
+    Thread.sleep(2 * size);
+    if (failure.get() != null) {
+      throw new Exception(failure.get());
+    }
```

Comment:
Wondering if this can be in a loop so the test can fail fast.
```
for (int i = 0; i < size / 100; i++) {
  Thread.sleep(100);
  if (failure.get() != null) {
    LOG.error(failure.get());
    fail(....);
  }
}